### PR TITLE
Added interval collapse implementation

### PR DESCRIPTION
**Pull request by c-monkey**
This fixes a bug discovered (Interval never implemented Collapse) resolving https://oncprojectracking.healthit.gov/support/browse/BONNIE-331

### CODE REVIEW

File: Src/coffeescript/cql-execution/src/elm/interval.coffee [hunk 1]
```diff
@@ -133,4 +133,31 @@ module.exports.Starts = class Starts extends UnimplementedExpression
...
+        if b.low <= a.high
+          a.high = b.high if b.high > a.high
+        else
+          intervals.push a
+          a = b
+        b = result.shift()
+      intervals.push a
+      result = intervals
+
+    result
```
Comment (edeyoung) [inline Src/coffeescript/cql-execution/src/elm/interval.coffee]: consider changing the name of `result` as the first variable. you call the arguments `result`, implying they will be returned. Then, at the end, you assign `intervals` to result, which is actually what's returned. This is confusing. Perhaps start with `intervals` and then add things to `collapsed_intervals` and then return `collapsed_intervals`.

Comment (c-monkey) [inline Src/coffeescript/cql-execution/src/elm/interval.coffee]: Nice suggestion. I've used camel case which seems like the norm for this file.

File: Src/coffeescript/cql-execution/src/elm/interval.coffee [hunk 2]
```diff
@@ -133,4 +133,36 @@ module.exports.Starts = class Starts extends UnimplementedExpression
...
+    if result?.length > 1
+      # we don't handle imprecise intervals at this time
+      for a in result
+        if a.low.isImprecise?() || a.high.isImprecise?()
+          throw new Error("Collapse does not support imprecise dates at this time.")
+
+      # sort intervals by start
+      result.sort (a,b)->
+        return -1 if a.low < b.low
+        return 1 if a.low > b.low
```
Comment (cmoesel) [inline Src/coffeescript/cql-execution/src/elm/interval.coffee]: If `a.low` is an instance of `DateTime`, I don't think you can use native `<` and `>` here.  You probably have to do something like:
```
if typeof a.low.before == 'function'
  return -1 if a.low.before b.low
  return 1 if a.low.after b.low
else
  return -1 if a.low < b.low
  return 1 if a.low > b.low
0
```

File: Src/coffeescript/cql-execution/src/elm/interval.coffee [hunk 3]
```diff
@@ -133,4 +133,36 @@ module.exports.Starts = class Starts extends UnimplementedExpression
...
+        0
+
+      # collapse intervals as necessary
+      collapsedIntervals = result
+      result = []
+      a = collapsedIntervals.shift()
+      b = collapsedIntervals.shift()
+      while b
+        if b.low <= a.high
```
Comment (cmoesel) [inline Src/coffeescript/cql-execution/src/elm/interval.coffee]: Again, those native `<=` operators shouldn't work if this is a `DateTime`.  For `<=` you'd have to use `sameOrBefore`.  Not sure why your tests are passing -- but it's probably just luck (or you're passing in the intervals in the right order already).

File: Src/coffeescript/cql-execution/src/elm/interval.coffee [hunk 4]
```diff
@@ -133,4 +133,36 @@ module.exports.Starts = class Starts extends UnimplementedExpression
...
+          a.high = b.high if b.high > a.high
```
Comment (cmoesel) [inline Src/coffeescript/cql-execution/src/elm/interval.coffee]: For `DateTime`, use `after` instead of `>`.

**Comment by cmoesel** [comment]
Code looks good, unit tests pass.  Did not test in Bonnie.

Will probably need to be rebased since another PR was just merged.

**Comment by c-monkey** [comment]
This is in Bonnie, we should merge the PR now. @cmoesel approves merging this into cql4bonnie branch.

**Comment by cmoesel** [comment]
I'll do it.

**Pull request merged**